cmd/bnscli/cmd_username.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/iov-one/weave"
+	weaveapp "github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/cmd/bnsd/app"
+	"github.com/iov-one/weave/x/username"
+)
+
+func cmdRegisterUsername(input io.Reader, output io.Writer, args []string) error {
+	fl := flag.NewFlagSet("", flag.ExitOnError)
+	fl.Usage = func() {
+		fmt.Fprintln(flag.CommandLine.Output(), `
+Create a transaction for registering a username.
+		`)
+		fl.PrintDefaults()
+	}
+	var (
+		nameFl       = fl.String("name", "", "Name part of the username. For example 'alice'")
+		namespaceFl  = fl.String("ns", "iov", "Namespace (domain) part of the username. For example 'iov'")
+		blockchainFl = fl.String("bc", "", "Blockchain network ID.")
+		addressFl    = flHex(fl, "addr", "", "Hex encoded blochain address on this network.")
+	)
+	fl.Parse(args)
+
+	uname, err := username.ParseUsername(*nameFl + "*" + *namespaceFl)
+	if err != nil {
+		return fmt.Errorf("given data produce an invalid username: %s", err)
+	}
+
+	msg := username.RegisterTokenMsg{
+		Metadata: &weave.Metadata{Schema: 1},
+		Username: uname,
+		Target: username.Location{
+			BlockchainID: *blockchainFl,
+			Address:      *addressFl,
+		},
+	}
+	if err := msg.Validate(); err != nil {
+		return fmt.Errorf("given data produce an invalid message: %s", err)
+	}
+
+	tx := &app.Tx{
+		Sum: &app.Tx_RegisterTokenMsg{
+			RegisterTokenMsg: &msg,
+		},
+	}
+	_, err = writeTx(output, tx)
+	return err
+}
+
+func cmdResolveUsername(input io.Reader, output io.Writer, args []string) error {
+	fl := flag.NewFlagSet("", flag.ExitOnError)
+	fl.Usage = func() {
+		fmt.Fprintln(flag.CommandLine.Output(), `
+Query a node to resolve the username. Successful result outputs a JSON
+serialized representation of the resolved username.
+		`)
+		fl.PrintDefaults()
+	}
+	var (
+		tmAddrFl = fl.String("tm", env("BNSCLI_TM_ADDR", "https://bns.NETWORK.iov.one:443"),
+			"Tendermint node address. Use proper NETWORK name. You can use BNSCLI_TM_ADDR environment variable to set it.")
+		nameFl      = fl.String("name", "", "Name part of the username. For example 'alice'")
+		namespaceFl = fl.String("ns", "iov", "Namespace (domain) part of the username. For example 'iov'")
+	)
+	fl.Parse(args)
+
+	uname, err := username.ParseUsername(*nameFl + "*" + *namespaceFl)
+	if err != nil {
+		return fmt.Errorf("given data produce an invalid username: %s", err)
+	}
+
+	token, err := fetchUsernameToken(*tmAddrFl, uname)
+	if err != nil {
+		return fmt.Errorf("cannot fetch token: %s", err)
+	}
+
+	raw, err := json.MarshalIndent(token, "", "\t")
+	if err != nil {
+		return fmt.Errorf("cannot json serialize token information: %s", err)
+	}
+	_, err = output.Write(raw)
+	return err
+}
+
+func fetchUsernameToken(serverURL string, uname username.Username) (*username.Token, error) {
```

Comment:
This is a lot of detail work. We should add better helpers for queries.

This is not a comment on the pr, but on the cli in general..

Reply:
And I think it is my open issue since 2 months